### DISCUSSION TITLE
remark-disable-tokenizers not working with remark 13, options?

### DISCUSSION THREAD

**Discussion by janpe** [remarkjs/remark#757] [Q&A] [answered]
`remark-disable-tokenizers` seems not to work with remark 13. Are there currently any options for reaching the same result?

I'd like to have
```
- List
- Of
- Items
```

to be output as text unchanged instead of making it a list.

**Comment by wooorm** [comment, marked as answer]
Looks like some of the markdown in your question is not working: See the [CommonMark Help](https://commonmark.org/help/) page for more on markdown.

As it’s a community plugin, it might be good to ask and/or solve the problem there? Notably, there are ways to turn off constructs indeed.

As your question is explained in only a few sentences, it might also be an [XY problem](https://xyproblem.info). If you could spend more time framing your question, we might be able to offer a better solution.

**Reply by janpe** [reply]
Sorry, fixed the markdown in my question.

Thanks for the XY problem link. I'll try to explain it better.

So I am using `react-markdown` and I want content to be parsed as markdown, but I want some exceptions. I want content that would be parsed as ordered lists and unordered lists to be returned without changes. So I want the output of

```
- One 
- Two
```

to be just that as text instead of being rendered as inside `ul` and `li` elements.

I started discussing this issue in the `react-markdown` repository but got the suggestion to ask here since it is not a question specific to `react-markdown`. The community plugin in question seems not too actively maintained.

**Reply by wooorm** [reply]
- What other exceptions do you want?
- Why don’t you want lists?

> The community plugin in question seems not too actively maintained.

I don’t think that’s completely true, the project in question receives regular updates and in my experience the folks working on it often respond relatively quickly.
I am under the impression that they would appreciate a PR to update `remark-disable-tokenizers`.

**Reply by janpe** [reply]
The content I get fetch for my client sometimes begins a paragraph with character that would be parsed as a list and I want to avoid that and render the text as is.

I'm not too familiar with remark and its plugins, I just stumbled into them working with my current issue. I checked out the code in the plugin and wasn't able to figure out what the required changes are before trying to ask for help.

I don't have all the time I want to use on this issue as I believe you know a developer sometimes doesn't. But if there is a good resource on migrating the plugins, I'd be glad to see if I can manage to find the time to contribute.

Thanks and sorry for not presenting my issue in the best way possible.

**Reply by wooorm** [reply]
It sounds like the solution might be to teach your users how markdown works, or show how it renders earlier. If otherwise you say you use markdown, but don’t support lists, that might break their expectation.

> I don't have all the time I want to use on this issue as I believe you know a developer sometimes doesn't. 

That’s fine, you get to decide how you spend your time. But also please understand that I enjoy writing code and sharing it with the world, and not necessarily to try and help someone who doesn’t have the time talk about their problem.

> I'm not too familiar with remark and its plugins, I just stumbled into them working with my current issue. I checked out the code in the plugin and wasn't able to figure out what the required changes are before trying to ask for help.

With the changelog for remark 13 in hand, some time spent learning [what all these projects do](https://unifiedjs.com) is and how [micromark extensions](https://github.com/micromark/micromark#extensions) work, it should be possible to figure out, and if not, feel free to ask questions.

**Reply by janpe** [reply]
> It sounds like the solution might be to teach your users how markdown works, or show how it renders earlier. If otherwise you say you use markdown, but don’t support lists, that might break their expectation.

I understand this and would prefer it to be as you described but in this particular case I have no way of doing this. 

> That’s fine, you get to decide how you spend your time. But also please understand that I enjoy writing code and sharing it with the world, and not necessarily to try and help someone who doesn’t have the time talk about their problem.

I have the time to talk about it, that's not the problem. As I said previously, I apologise for not providing enough information at first.
 
> With the changelog for remark 13 in hand, some time spent learning [what all these projects do](https://unifiedjs.com) is and how [micromark extensions](https://github.com/micromark/micromark#extensions) work, it should be possible to figure out, and if not, feel free to ask questions.

I appreciate the help. I'll see if I manage to do something about it.

I guess there's no reason to spend any more time on this discussion since the solution is either to influence the data that is coming to the client (which I cannot do) or for me to contribute to the plugin.

Thanks for your time, it's very much appreciated.

**Reply by ChristianMurphy** [reply]
> solution is either to influence the data that is coming to the client (which I cannot do)

You're using `react-markdown`, usually that is used in an authoring experience.
In which case, yes, you do have the ability to influence over what is coming from the client.

Reading between the lines a bit there may be legacy content which expects some markdown syntax to be treated as plain text.
That could be migrated to plain markdown, by creating a plugin, to escape constructs which previously rendered as plain text, see https://astexplorer.net/#/gist/1898fb07a2e8e6b80e5a8ba403b9dd52/3752ac9092b70528db993eedf19a45dd1f97ca47 and https://unifiedjs.com/learn then saving the content back to where it is stored, and using markdown going forward.

**Reply by janpe** [reply]
Thanks Christian. I hope you can respect the fact that I have to work with the constraints that I have described. In the long run a more robust system is possible but right here and now I need to manage the situation and I prefer using a plugin that works with the newer version rather than relying on an older version of `react-markdown`, which also similarly requires the `remark-disable-tokenizers` plugin to achieve what I need.

Thanks for providing some ideas, it has helped me figure out a way to handle the situation as it is right now.

**Reply by WCByrne** [reply]
@janpe Did you ever make any progress on this? I'm in the same situation but thought I'd check in before jumping in to trying to learn how to update the plugin.

**Reply by janpe** [reply]
@WCByrne I did manage to create a tiny plugin as per the comments on this issue: https://github.com/zestedesavoir/zmarkdown/issues/416#issuecomment-982812961

**Reply by WCByrne** [reply]
So I got a little "plugin" working but it seems like I can't disable some things. Notably images. ~Is there a definitive list of values that can be disabled?~ Found this [list of constructs](https://github.com/micromark/micromark/blob/b866e34551cf3ad34250c58e92f322714ba9d8da/packages/micromark/dev/lib/constructs.js)

Also, I need to support strikethrough which only appears to be part of GFM, but when I do that it seems I can't disable any of the other  GFM additions like tables. 

I feel like I'm starting to spin circles trying to navigate the documentation on all of this.

**Comment by ChristianMurphy** [comment]
Taking a step back for a moment.
Why are you looking for something markdown-like, instead of using markdown?
Where is the content coming from? (if not from react-markdown or remark, why is that system not generating the HTML?)
Context helps a lot in answering questions. :slightly_smiling_face: 

---

> `remark-disable-tokenizers` seems not to work with remark 13. Are there currently any options for reaching the same result?

One possibility, would be to address https://github.com/zestedesavoir/zmarkdown/issues/416, allowing the plugin to work with remark 13.

**Comment by samdenty** [comment]
for anyone looking for a solution to `remark-disable-tokenizers` in remark 13+:

```ts
import { SKIP, visit } from 'unist-util-visit';
import type { Plugin } from 'unified';
import type { UnistNode, UnistParent } from 'node_modules/unist-util-visit/lib';

const limitedMarkdownPlugin: Plugin = () => {
  return (tree, file) => {
    const contents = file.toString();

    visit(tree, (node: UnistNode, index, parent: UnistParent) => {
      if (
        index == null ||
        ['paragraph', 'text', 'inlineCode', 'code', 'strong', 'emphasis'].includes(node.type) ||
        !node.position
      ) {
        return true;
      }

      let value = contents.slice(node.position.start.offset, node.position.end.offset);

      if (node.type === 'heading') {
        value = `\n${value}`;
      }

      parent.children[index] = {
        type: 'text',
        value,
      } as any;

      return [SKIP, index] as const;
    });
  };
};
```

**Reply by wooorm** [reply]
That looks fragile. I a) recommend not doing this, b) recommend going with https://github.com/micromark/micromark#case-turn-off-constructs
